### PR TITLE
loader: add support for initrd

### LINKED ISSUE #208: Add Support for initrd_path
With the initial API when specifying a kernel path, you could also specify the path of the initrd file.
For now this feature is not mandatory.
Acceptance Criteria: Decide if this feature is needed for other use cases.

### PR DESCRIPTION
**Pull request by marcov**
## Reason for This PR

The [2020 Roadmap](https://github.com/firecracker-microvm/firecracker/issues/1104) welcomes contributions that adds initrd support. A previous PR was opened to add it (#670), but cannot be reused as is, because it also added support for loading a compressed kernel.

Hence #670 was reworked and rebased to only include initrd support.
Fixes #208 .

## Description of Changes

- Add `initrd_path` property to the `boot-source` API.
- Pass `ramdisk_image` (memory offset) and `ramdisk_size` to the guest bootparams table.
- Implement the `load_initrd` loader function, to load the initrd file content into the guest memory ~~at a (for the moment) fixed address~~ at then end of the available low memory (for x86_64).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.

### CODE REVIEW

**Comment by marcov** [comment]
Disclaimer: I am complete rust n00b!

Missing things:
- [ ] aarch64 support
- [ ] tests (#670 had dedicated commits for tests that can be ported too).

**Comment by marcov** [comment]
For x86_64 I am now loading the initrd at the end of the available low memory, instead of using a fixed address.

**Comment by marcov** [comment]
Hi, x86_64 is now ready for review. I just need a new fixture added to S3 for initrd testing.

Do you expect aarch64 be supported too, or can it be postponed in a later PR?

**Comment by rgooch** [comment]
I would suggest merging and postponing aarch64, so that x86_64 people can test sooner (more easily).

**Comment by marcov** [comment]
Spent some more hours and added `aarch64` support. But I don't have any `aarch64` hardware with KVM hypervisor mode to be able to test it.

**Comment by marcov** [comment]
Update: aarch64 is functional. 
Also, almost all tests are passing, just waiting for a initrd update on S3 to align the ramdisk images with the tests.

**Comment by marcov** [comment]
All tests are now passing apart `test_microvm_initrd_with_serial` on aarch64
It needs a kernel update on S3, since the current kernel does not support booting with initrd:
```
# grep initrd /proc/kallsyms
#
```

File: api_server/src/request/boot_source.rs [hunk 1]
```diff
@@ -29,13 +29,15 @@ mod tests {
     use super::*;
 
     #[test]
-    fn test_into_parsed_request() {
+    fn test_into_parsed_request_same() {
```
Comment (andreeaflorescu) [inline api_server/src/request/boot_source.rs]: I don't think we need two tests here. We can have just one where initrd_path has a value.

File: api_server/src/request/boot_source.rs [hunk 2]
```diff
@@ -46,4 +48,25 @@ mod tests {
...
+            initrd_path: None,
+            boot_args: Some(String::from("foobar")),
+        };
+        let initrd_body = BootSourceConfig {
+            kernel_image_path: String::from("/foo/bar"),
+            initrd_path: Some(String::from("/bar/baz")),
+            boot_args: Some(String::from("foobar")),
+        };
+        let (sender, receiver) = oneshot::channel();
+        assert!(!body
```
Comment (andreeaflorescu) [inline api_server/src/request/boot_source.rs]: This assert doesn't look good. If you look at the previous test, it checks that body.into_parsed_request equals ConfigureBootSource(same_body, sender). In that test body and same_body are the same thing. We created two object with the same data because `BootSourceConfig` does not implement copy so basically we just create two object with same data to test the "equality".

File: memory_model/src/guest_memory.rs [hunk 3]
```diff
@@ -139,6 +139,15 @@ impl GuestMemory {
         self.regions.len()
     }
 
+    /// Returns the size region identified by index
+    pub fn region_size(&self, index: usize) -> usize {
+        if index >= self.regions.len() {
```
Comment (andreeaflorescu) [inline memory_model/src/guest_memory.rs]: In this case I think it's better to return an error.

File: tests/integration_tests/performance/test_boottime.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
 # The maximum acceptable boot time in us.
 MAX_BOOT_TIME_US = 150000
+MAX_BOOT_TIME_INITRD_US = 400000
```
Comment (andreeaflorescu) [inline tests/integration_tests/performance/test_boottime.py]: Hmm, I was expecting the difference to be smaller between with initrd_path and without. I wonder if that difference is expected?

File: tests/integration_tests/build/test_licenses.py [hunk 5]
```diff
@@ -10,22 +10,25 @@
...
+)
 CHROMIUM_COPYRIGHT = (
     "Copyright 2017 The Chromium OS Authors. All rights reserved."
-    )
+)
 CHROMIUM_LICENSE = (
     "Use of this source code is governed by a BSD-style license that can be"
-    )
+)
+SUSE_COPYRIGHT = (
```
Comment (andreeaflorescu) [inline tests/integration_tests/build/test_licenses.py]: Where is the SUSE copyright added?

Comment (marcov) [reply tests/integration_tests/build/test_licenses.py]: I added `tests/integration_tests/functional/test_initrd.py`. If I can put the initrd tests in an already copyrighted file,  then this is not needed.

File: tests/integration_tests/performance/test_boottime.py (hunk 4 above)
Comment (marcov) [reply tests/integration_tests/performance/test_boottime.py]: That value is definitely too high. From the last test I get:

```
integration_tests/performance/test_boottime.py 
Boot time with no network is: 99778 us
...Boot time with network configured is: 118604 us
.Boot time with initrd + network configured is: 148323 us
.Boot time with initrd is: 150110 us
```

- initrd tests a.t.m. are always run with a network enabled, so there's the 20ms overhead from that,
- the remaining 30ms should be due to initrd specific code (checking for cmpression method, mounting the rootfs partition,...).

Comment (andreeaflorescu) [reply tests/integration_tests/performance/test_boottime.py]: Okay, so basically we could safely use 160000 us instead?

File: memory_model/src/guest_memory.rs (hunk 3 above)
Comment (marcov) [reply memory_model/src/guest_memory.rs]: It is returning an `Result<usize>` in the latest commit, I have messed up with commit orders / squashing

File: api_server/src/request/boot_source.rs (hunk 1 above)
Comment (marcov) [reply api_server/src/request/boot_source.rs]: Ok!

File: api_server/src/request/boot_source.rs (hunk 2 above)
Comment (marcov) [reply api_server/src/request/boot_source.rs]: Understood, test updated, thanks!

File: tests/integration_tests/performance/test_boottime.py (hunk 4 above)
Comment (marcov) [reply tests/integration_tests/performance/test_boottime.py]: If you think 10ms accounts for the boot time variance, then yes!

**Comment by marcov** [comment]
@andreeaflorescu, kind ping :-)

Let me know if you are waiting for any action from my side to finalize this PR.

**Comment by andreeaflorescu** [comment]
@marcov it looks like the tests are failing on both x86_64 and aarch64. I think it is because of a rebase: https://github.com/firecracker-microvm/firecracker/blob/695fd7fee535f219e6db618c1302ef2e98b0b48d/vmm/src/lib.rs#L1959. Can you fix the failing tests? You will need 2 reviewers for you PR. I'll try to find someone tomorrow.

File: tests/integration_tests/functional/test_initrd.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+"""Tests for initd."""
+
+import pytest
+from host_tools.network import SSHConnection
+from framework.microvm import Serial
+
+INITRD_FILESYSTEM = "rootfs"
+
+
+@pytest.mark.xfail(raises=ConnectionError)
```
Comment (andreeaflorescu) [inline tests/integration_tests/functional/test_initrd.py]: Why do you expect this test to fail?

File: tests/integration_tests/functional/test_initrd.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    vm.basic_config(
+        add_root_device=False,
+        boot_args='random.trust_cpu=on console=ttyS0 reboot=k panic=1 pci=off',
+        use_initrd=True
+    )
+
+    vm.start()
+    serial = Serial(vm)
+    serial.open()
+    serial.rx(token='login: ')
```
Comment (andreeaflorescu) [inline tests/integration_tests/functional/test_initrd.py]: This assert seems to be failing when testing on aarch64. Any reason why the serial would not work with the initrd you provided? Do you happen to know why is it failing? Is there something that we have to change to the resources?

Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: That's because the aarch64 kernel does not support initrd loading. You should update it on S3.

Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: That's because the aarch64 kernel does not support initrd loading. You should update it on S3.

File: tests/integration_tests/functional/test_initrd.py (hunk 6 above)
Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: I have the SSH connection to the guest fail depending on the kernel / machine used to test. 
What I could see it's that it takes a while for the SSH server to come up, and I suspect it may be because the random source is not generating fast enough. There was a discussion about that on the Slack channel about 1 month ago.

File: tests/integration_tests/functional/test_initrd.py (hunk 7 above)
Comment (andreeaflorescu) [reply tests/integration_tests/functional/test_initrd.py]: I just checked yesterday with @dhrgit and he says  he uploaded a kernel with support for initrd for aarch64 as well. I also looked at the S3 bucket and there's minimal_with_initrd also for aarch64. Can it be that the serial does not work for another reason? I'll check it tomorrow myself as well. We are attending a meetup these days, sorry for the late reply.

File: tests/integration_tests/functional/test_initrd.py (hunk 6 above)
Comment (andreeaflorescu) [reply tests/integration_tests/functional/test_initrd.py]: Can we just increase the timeout instead? It might be that the tests never get to run which I don't think it's the best solution.

File: tests/integration_tests/functional/test_initrd.py (hunk 7 above)
Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: Thanks for the info. I assumed the initrd support is missing in kernel because (one month I ago) I did not get any match for `grep initrd /proc/kallsyms` on a aarch64 VM booted with rootfs. If you can check that's great, otherwise I'll try to spend some more time on it.

File: tests/integration_tests/functional/test_initrd.py (hunk 6 above)
Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: Let me do some more tests and I'll tell you

File: tests/conftest.py [hunk 8]
```diff
@@ -116,6 +116,10 @@ def test_with_any_microvm(test_microvm_any):
 session root and temporary directories.
```
Comment (dianpopa) [inline tests/conftest.py]: @marcov, this is commit is nice functionality. I encourage you to make a different PR containing just this commit since it s totally decoupled from the initrd support and it has higher chances of getting merged faster.

**Comment by dianpopa** [comment]
I tried to replicate the failure on aarch64 for the `test_microvm_initrd_with_serial` and indeed the kernel panics with message:
`
[    0.204197] VFS: Cannot open root device "(null)" or unknown-block(0,0): error -6
[    0.206833] Please append a correct "root=" boot option; here are the available partitions:
[    0.209752] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    0.212721] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.14.138 #1
[    0.214940] Hardware name: linux,dummy-virt (DT)

`
As far as my previous knowledge, this could also mean that the `initrd.img` is not correctly formatted.
How exactly were the `initrd.img`created for both x86_64 and aarch64?

File: tests/conftest.py (hunk 8 above)
Comment (marcov) [reply tests/conftest.py]: Sure!

**Comment by marcov** [comment]
@dianpopa: the initrd is built using https://github.com/marcov/firecracker-initrd

*If I recall correctly*, I did test the same initrd with a different aarch64 kernel image and it was working.

**Comment by dianpopa** [comment]
> @dianpopa: the initrd is built using https://github.com/marcov/firecracker-initrd
> 
> _If I recall correctly_, I did test the same initrd with a different aarch64 kernel image and it was working.

Thanks, I will try create an initrd for aarch64 myself and build a kernel with initrd. I will let you know the results. If I am successful I will also update the `minimal_with_initrd` folder from S3.

File: tests/integration_tests/functional/test_initrd.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+import pytest
+from host_tools.network import SSHConnection
+from framework.microvm import Serial
+
+INITRD_FILESYSTEM = "rootfs"
+
+
+@pytest.mark.xfail(raises=ConnectionError)
+def test_microvm_initrd_with_ssh(
```
Comment (dianpopa) [inline tests/integration_tests/functional/test_initrd.py]: I see that the initrd provided does not add support for networking, the eth0 interface cannot be found by the networking service upon guest start.  So this could explain why this test is failing. However, as far as my concern, this test could be deleted since initd support should work without network support.

Comment (marcov) [reply tests/integration_tests/functional/test_initrd.py]: Deleted

**Comment by marcov** [comment]
> @marcov please solve the conflicts that github is complaining about so that I can continue reviewing the latest code changes.
> 
> * I uploaded a new kernel to S3 for aarch64 that should pass the tests for aarch64 too

Thank you @dianpopa, aarch64 kernel is now OK.

> * I want us to upload the kernel configs used for building kernels with initrd support in the "resources" folder of the codebase. I already have them, so if it is ok with you I can make a different commit on top of your commits after you solve conflicts. let me know.

Sure, conflicts are now solved.

**Comment by marcov** [comment]
Also, I'd like to see this reviewed soon, rebasing on master was harder than expected because of the many changes. Thank you!

**Comment by dianpopa** [comment]
> Also, I'd like to see this reviewed soon, rebasing on master was harder than expected because of the many changes. Thank you!

We will do our best to review this asap!

File: kernel/src/loader/mod.rs [hunk 10]
```diff
@@ -258,6 +264,87 @@ pub fn load_cmdline(
...
+        Ok(s) => initrd_size = s as usize,
+    };
+
+    let align_to_pagesize = |address| address & !(PAGE_SIZE - 1);
+
+    initrd_image
+        .seek(SeekFrom::Start(0))
+        .map_err(|_| Error::ReadInitrd)?;
+
+    let region_size: usize = guest_mem.region_size(0).map_err(|_| Error::LoadInitrd)?;
```
Comment (dianpopa) [inline kernel/src/loader/mod.rs]: What is the reason for loading the initrd at the end of the first memory region? I may be missing something but in order to get the end of the low memory you should do: 
```
guest_mem.end_addr()
```

File: kernel/src/loader/mod.rs [hunk 11]
```diff
@@ -17,6 +17,8 @@ use super::cmdline::Error as CmdlineError;
 use memory_model::{Address, GuestAddress, GuestMemory};
 use sys_util;
 
+const PAGE_SIZE: usize = 4096;
```
Comment (dianpopa) [inline kernel/src/loader/mod.rs]: My understanding is that you are trying to round up to the smallest page size possible, right? If that is so, why not add a comment above, something like:
```
// Smallest page size for the supported architectures.
const PAGE_SIZE: usize = 4096;
```

File: arch/src/aarch64/fdt.rs [hunk 12]
```diff
@@ -113,7 +113,12 @@ pub fn create_fdt<T: DeviceInfoForFDT + Clone + Debug>(
     append_property_u32(&mut fdt, "interrupt-parent", GIC_PHANDLE)?;
```
Comment (dianpopa) [inline arch/src/aarch64/fdt.rs]: This commit needs to be squashed into the first commit.

File: kernel/src/loader/mod.rs (hunk 10 above)
Comment (marcov) [reply kernel/src/loader/mod.rs]: You are right, I need the "end" of the low memory. I though `end_addr()` would return addressed above 4G.

File: kernel/src/loader/mod.rs (hunk 11 above)
Comment (marcov) [reply kernel/src/loader/mod.rs]: Correct, I'll add that comment

File: kernel/src/loader/mod.rs (hunk 10 above)
Comment (marcov) [reply kernel/src/loader/mod.rs]: @dianpopa, actually here I need the size of the low memory region, not the end address, since I am assuming the region may not start at offset 0.

**Comment by marcov** [comment]
Pushed a new version with changes based on reviews.

File: kernel/Cargo.toml [hunk 13]
```diff
@@ -3,5 +3,6 @@ name = "kernel"
 version = "0.1.0"
 
 [dependencies]
+arch = { path = "../arch" }
```
[the remark this answers is not in the document]
Comment (marcov) [reply kernel/Cargo.toml]: BTW, I really appreciate you taking the time to explain me what was not right in my code, and provide a better solution! I wish all reviews were this well made ;-)

**Comment by dianpopa** [comment]
> Pushed a new version with changes based on reviews.

Cool, Ill review your latest changes while also trying to see why the aarch64 test fail.

File: kernel/Cargo.toml (hunk 13 above)
[the remark this answers is not in the document]
Comment (dianpopa) [reply kernel/Cargo.toml]: > BTW, I really appreciate you taking the time to explain me what was not right in my code, and provide a better solution! I wish all reviews were this well made ;-)

Nice, to hear this! Thanks for the feedback :)

**Comment by marcov** [comment]
About the aarch64 failure, the reason is that in test_boottime the `Guest-boot-time` line takes a while before being printed, and the `time.sleep()` call may not wait enough. 
A way to avoid that is to keep reading the log with a timeout, until the line the line is found.

**Comment by dianpopa** [comment]
> About the aarch64 failure, the reason is that in test_boottime the `Guest-boot-time` line takes a while before being printed, and the `time.sleep()` call may not wait enough.
> A way to avoid that is to keep reading the log with a timeout, until the line the line is found.

After a little bit of debugging, I noticed the the "Guest-boot-time" message is not outputted at all. Seems that this [mmap](https://github.com/marcov/firecracker-initrd/blob/master/guest/boot_done.c#L18) fails :(. @marcov , as far as I remember the tests were passing after I uploaded a new kernel to S3 but I cannot recall whether or not you did have the initrd boottime test back then. I am asking this because I am trying to figure out what changed in the meantime.

**Comment by marcov** [comment]
The boot time performance tests were enabled just recently for aarch64, so maybe that's a reason why we did not see the failure before?

Anyway, when testing the initrd kernel with a rootfs the boot-time log is generated, so there's likely something wrong in the inirtd init launcher.

**Comment by dianpopa** [comment]
> The boot time performance tests were enabled just recently for aarch64, so maybe that's a reason why we did not see the failure before?
> 
> Anyway, when testing the initrd kernel with a rootfs the boot-time log is generated, so there's likely something wrong in the inirtd init launcher.
> 
> In addition, I also tried to create a custom initrd by using as init with the `/bin/busybox` found in a rootfs image. When running it, I am getting: `Can't map memory: Bad file descriptor`

I am able to replicate the failure (meaning I do not receive the log message) by using the `./firecracker/build/img/aarch64/minimal_with_initrd/kernel/vmlinux.bin` as kernel path and the `./firecracker/build/img/aarch64/minimal_with_initrd/fsfiles/initrd.img` as initrd_path which is what the [failing test](https://github.com/firecracker-microvm/firecracker/pull/1246/files#diff-5996dacca5f8a1a7eed702c084674b73R101) is actually using and leaving behind. But as far as I understand, you are able to see a boot-time log when you do the same, right?
I also tried using the `./firecracker/build/img/aarch64/minimal/fsfiles/boottime-rootfs.ext4` to create an initrd but the kernel complains that it cannot unpack it and I have no idea why.

**Comment by marcov** [comment]
>  But as far as I understand, you are able to see a boot-time log when you do the same, right?

Sorry, no, I can also reproduce the problem locally for aarch64.

Basically, the mmio write is not working as expected. 

> I also tried using the `./firecracker/build/img/aarch64/minimal/fsfiles/boottime-rootfs.ext4` to create an initrd but the kernel complains that it cannot unpack it and I have no idea why.

initrd should be be an ASCII cpio archive. 

Another test worth doing is to create a custom rootfs with [this init](https://github.com/marcov/firecracker-initrd/blob/master/guest/boot_done.c) and see if the boot message log is generated.

**Comment by marcov** [comment]
Ok, I think I found an explanation. 
The mmap fails because the fd for `/dev/mem` is bad (I should have checked that on open(), my fault), and that's because the `/dev/` mount point is not yet set up when the `open()` is done.

**Comment by dianpopa** [comment]
> Ok, I think I found an explanation.
> The mmap fails because the fd for `/dev/mem` is bad (I should have checked that on open(), my fault), and that's because the `/dev/` mount point is not yet set up when the `open()` is done.

Cool! I was also suspecting that it might be something wrong with `dev/mem` especially because if you do a `ls -al /dev/` after booting the alpine image with an initrd the devices have wrong dates (some even have the classic 1 JAN 1970). Maybe our problem  might be related to that.

**Comment by marcov** [comment]
Hi @dianpopa, I pushed a initrd fix for the issue https://github.com/marcov/firecracker-initrd/commit/d2641e3cb88c797a1ecbefee5094f1fd4b0a6360 (I can finally get the boot done log with that!)
Though I am not sure if the performance test is going to stay within the maximum boot time...

When you have some time, can you please rebuild the initrd and put it on S3?

**Comment by dianpopa** [comment]
> Hi @dianpopa, I pushed a initrd fix for the issue [marcov/firecracker-initrd@d2641e3](https://github.com/marcov/firecracker-initrd/commit/d2641e3cb88c797a1ecbefee5094f1fd4b0a6360) (I can finally get the boot done log with that!)
> Though I am not sure if the performance test is going to stay within the maximum boot time...
> 
> When you have some time, can you please rebuild the initrd and put it on S3?

Oh, nice! Thanks for the fix, I will upload it in S3 soon enough.

File: src/vmm/src/lib.rs [hunk 14]
```diff
@@ -1005,6 +1006,44 @@ impl Vmm {
...
+    ) -> std::result::Result<InitrdInfo, LoadInitrdError>
+    where
+        F: Read + Seek,
+    {
+        use LoadInitrdError::*;
+
+        let size: usize;
+        // Get the image size
+        match image.seek(SeekFrom::End(0)) {
+            Ok(0) | Err(..) => return Err(ReadInitrd),
```
Comment (andreeaflorescu) [inline src/vmm/src/lib.rs]: We shouldn't ignore the error here. What we can do instead is return ReadInitrd(EINVAL) when seek returns 0 and the seek error on the Err(..) branch.

**Comment by andreeaflorescu** [comment]
@marcov do you need help in addressing the outstanding review comments?

File: src/vmm/src/lib.rs [hunk 15]
```diff
@@ -1005,6 +1006,44 @@ impl Vmm {
         Ok(entry_addr)
     }
 
+    /// Loads the initrd from a file into the given memory slice.
```
[the remark this answers is not in the document]
Comment (marcov) [reply src/vmm/src/lib.rs]: >  Is there a technical reason for which this is needed in vmm/src/lib.rs

This is because the load address of the initrd is computed in the arch module, and I wanted to avoid a dependency of the `kernel` module from `arch`.

File: src/arch/src/lib.rs [hunk 16]
```diff
@@ -50,6 +50,17 @@ pub enum DeviceType {
     RTC,
 }
 
+/// Type for passing information about the initrd in the guest memory.
```
[the remark this answers is not in the document]
Comment (marcov) [reply src/arch/src/lib.rs]: Again, because `arch` makes use of this type, and moving it in kernel would make `arch` depend on `kernel`.

**Comment by marcov** [comment]
Thank you for the ping @andreeaflorescu, pushed some updates now.

**Comment by andreeaflorescu** [comment]
@marcov I created a prototype on my branch on top of your commits: https://github.com/andreeaflorescu/firecracker/commit/994b23d43d2b7fd52cb8e3f9641b2c16f4a5b689.

The `initrd_load_addr` function doesn't need to live in arch. It is more of a utility function for loading the initrd image into guest memory. I moved it to the kernel crate along with `load_initrd` function. To break the dependency between the arch crate and the `kernel` crate, we can send the page size as a parameter to `load_initrd`. This is still not an ideal state because now arch depends on `kernel` which is not something that we want. This dependency can further be broke by refactoring the fdt implementation in arch, but this is out of scope of this PR.

To give you more context, it is important to have the loading of initrd in the `kernel` crate  because we are working towards consuming the `linux-loader` crate from rust-vmm. Once we make sure that loading the initrd works as expected, we should also send a PR with the code here: https://github.com/rust-vmm/linux-loader. If you don't want to also send a PR in rust-vmm, one of the Firecracker maintainers can take care of that. Loading an initrd image is quite a common feature that we would like to have in linux-loader as well. Having that functionality in the vmm crate makes it impossible to share the common code with the other VMMs written in Rust (for example with Cloud Hypervisor). Besides this, there is an ongoing effort to minimize the code in vmm.

**Comment by dianpopa** [comment]
> @marcov I created a prototype on my branch on top of your commits: [andreeaflorescu@994b23d](https://github.com/andreeaflorescu/firecracker/commit/994b23d43d2b7fd52cb8e3f9641b2c16f4a5b689).
> 
> The `initrd_load_addr` function doesn't need to live in arch. It is more of a utility function for loading the initrd image into guest memory. I moved it to the kernel crate along with `load_initrd` function. To break the dependency between the arch crate and the `kernel` crate, we can send the page size as a parameter to `load_initrd`. This is still not an ideal state because now arch depends on `kernel` which is not something that we want. This dependency can further be broke by refactoring the fdt implementation in arch, but this is out of scope of this PR.
> 
> To give you more context, it is important to have the loading of initrd in the `kernel` crate because we are working towards consuming the `linux-loader` crate from rust-vmm. Once we make sure that loading the initrd works as expected, we should also send a PR with the code here: https://github.com/rust-vmm/linux-loader. If you don't want to also send a PR in rust-vmm, one of the Firecracker maintainers can take care of that. Loading an initrd image is quite a common feature that we would like to have in linux-loader as well. Having that functionality in the vmm crate makes it impossible to share the common code with the other VMMs written in Rust (for example with Cloud Hypervisor). Besides this, there is an ongoing effort to minimize the code in vmm.

The `initrd_load_addr` **does need** to live in arch -> or at least this is how we achieve consistency with the way the kernel deals with initrd. Please take a look at how things work for x86_64 (starting here https://elixir.free-electrons.com/linux/v4.14.158/source/arch/x86/kernel/setup.c#L304) -> the first interaction of the kernel with the initrd config data is in the [setup_arch](https://elixir.free-electrons.com/linux/v4.14.158/source/arch/x86/kernel/setup.c#L850)  function which contains the [architecture-specific boot-time initializations](https://elixir.free-electrons.com/linux/v4.14.158/source/arch/x86/kernel/setup.c#L845) which is kind of the equivalent of our `configure_system` function. You can see there that the two essential things that the kernel needs are [the address](https://elixir.free-electrons.com/linux/v4.14.158/source/arch/x86/kernel/setup.c#L370) where the initrd is loaded and the [size](https://elixir.free-electrons.com/linux/v4.14.158/source/arch/x86/kernel/setup.c#L371) of it. 
* For the first one (i.e the address) you need to be aware of the memory layout for a specific platform (which we are defining in arch). The one function that does the job for that is `initrd_load_address` from arch. If we were to move this in the kernel crate (as it is the case with your model):
```
    let lowmem_size: usize = guest_mem.region_size(0).map_err(|_| Error::InitrdAddress)?;
```
we would make the kernel crate dependent on memory layout representation for a specific platform. Does not that break the separation of concerns we strive for?
* As far as the size, a fseek is sufficient. This functionality is achieved in the current implementation by the `load_initrd` function from the vmm.
After address and size is obtained, the image is written to memory. If we want to avoid passing around the guest memory (which is what we should be striving for in our decoupling goals), this should also stay in vmm.

**Comment by marcov** [comment]
Taking `load_kernel` as example, I can keep part of `load_initrd` in `vmm` to retrieve the initrd address and size, and have a `load_initrd` in `kernel` to write data into the guest memory. The `kernel` function would only be a one liner though.

**Comment by andreeaflorescu** [comment]
@marcov that was my initial thought as well and noticed the same problem as you. Let's keep it as is for now and we will find a better place for it in a future iteration.

I am still a bit concerned regarding the rather large boot time on aarch64. I was thinking about doing some comparison with  Qemu boot times (if they have something similar we can compare against). I can check that tomorrow.

**Comment by marcov** [comment]
@andreeaflorescu using a non-gzipped initrd will shave off around 100ms, but even doing that we are still far aways from the max time of 160ms. 
Checking the boot log timestamp, most of the time is spent by the kernel unpacking the initramfs, and that time depends on the size of the initrd file. This is with a 25MB initrd:
```
[    0.110152] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
[    0.112462] NET: Registered protocol family 1
[    0.114731] Unpacking initramfs...
[    0.353062] Freeing initrd memory: 24712K
```

**Comment by dianpopa** [comment]
> @andreeaflorescu using a non-gzipped initrd will shave off around 100ms, but even doing that we are still far aways from the max time of 160ms.
> Checking the boot log timestamp, most of the time is spent by the kernel unpacking the initramfs, and that time depends on the size of the initrd file. This is with a 25MB initrd:
> 
> ```
> [    0.110152] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
> [    0.112462] NET: Registered protocol family 1
> [    0.114731] Unpacking initramfs...
> [    0.353062] Freeing initrd memory: 24712K
> ```

On the a1.metal aarch64 machine (9 MB initrd), things look like this:
```
[    0.114070] Unpacking initramfs...
[    0.468239] Freeing initrd memory: 9420K
[    0.470117] audit: initializing netlink subsys (disabled)
[    0.472665] Initialise system trusted keyrings
```

**Comment by marcov** [comment]
@dianpopa  try to `gzip -d` the `initrd.img`, you should gain around 100ms.

**Comment by andreeaflorescu** [comment]
If it is expected for the boot time to be so large with the provided initrd image, we can merge the PR as is. I didn't really have time to investigate this, and I don't want to block the PR. Feel free to update the time for aarch64 to a value that you're confident with.

**Comment by dianpopa** [comment]
> If it is expected for the boot time to be so large with the provided initrd image, we can merge the PR as is. I didn't really have time to investigate this, and I don't want to block the PR. Feel free to update the time for aarch64 to a value that you're confident with.

+1 for this. With the above changes the, this PR should pass the CI:
```
```



> @dianpopa try to `gzip -d` the `initrd.img`, you should gain around 100ms.

I tried with the initrd.img that the CI leaves behind. @marcov, are you implying that it would be a good idea to reupload in S3 the gzipped version?

**Comment by dianpopa** [comment]
> @dianpopa try to `gzip -d` the `initrd.img`, you should gain around 100ms.

I tried with the initrd.img that the CI leaves behind. @marcov, are you implying that it would be a good idea to reupload in S3 the gzipped version?

**Comment by marcov** [comment]
@dianpopa: right! An `ASCII cpio archive` should be faster.

I'm going to do a final squash and set a reasonable boot time for aarch64, checking for `platform.machine()`

**Comment by marcov** [comment]
Happy to see this merged! 
Thanks for the help @dianpopa @andreeaflorescu , and happy holidays :-)

**Comment by dianpopa** [comment]
Thanks for the effort and responsiveness @marcov ! Happy holidays!
